### PR TITLE
GS/SW: Improve accuracy of line drawing and edge antialiasing.

### PR DESCRIPTION
**Pull request by TJnotJT**
### Description of Changes
1. Implement accurate rounding and pixel inclusion for lines and antialiased lines.
2. Implement improved pixel inclusion for antialiased triangle edges.

### Rationale behind Changes

Based on hardware tests this appears to give accurate pixel inclusion for lines, both aliased and antialiased. The actual value of coverage (aka alpha) is still not totally accurate; PS2 is doing some rounding/trunation to alpha that requires more research.

Similarly, pixel inclusion for antialiased edges of triangles is improved based on hardware tests. The inclusion is still not 100% accurate because there are some edge cases where the pixel can sometimes be +/-1 outside the bbox of the triangle that I haven't totally figured out.

As fars as games, the main fix is for Largo Winch (uses lines to draw the border color of shadow maps) and a FFXII FMV (appears to be similar). There are some small fixes to UI elements and such.

Note there are also some false positives for regressions. I have checked some of these and they either appear to occur on real hardware or are different issues (more details below).

### Suggested Testing Steps

So far I have done dump runs in SW for regression testing and several homebrew tests to compare accuracy with the actual hardware (can share the tests if there is an interest). For some dumps that seems to have regressions, I have checked them on a PS2 and confirmed that they look similar or differences are due to different issue (exampled provided below).

Testing on games know to use antialiasing or that require accurate lines would be helpful. Testing the performance vs. master would be good to make sure the performance impact is not too much.

### Did you use AI to help find, test, or implement this issue or feature?
I use Copilot for autocompletion.

### Improvements

**Largo Winch - Empire Under Threat_SLES-50841_20230120221136.gs.xz**

Master
S004569_07689_f00004_fr1_00000_C_32_Largo Winch - Empire Under Threat_SLES-50841_20230120221136 gs xz_master

PR
S004569_07689_f00004_fr1_00000_C_32_Largo Winch - Empire Under Threat_SLES-50841_20230120221136 gs xz_accline

**Final Fantasy XII preload frame data texture isses.gs.xz**

Master
S003707_00145_f00005_fr-1_00e00_C_32_Final Fantasy XII preload frame data texture isses gs xz_master

PR
S003707_00145_f00005_fr-1_00e00_C_32_Final Fantasy XII preload frame data texture isses gs xz_accline

### False positive regression/other issues

**Booting PS2 BIOS... _20220901235528.gs.xz**

Some corners of cubes stand out more in PR and PS2

Master
S003455_00476_f00001_fr1_00a00_C_32_Booting PS2 BIOS  _20220901235528 gs xz_master

PR
S003455_00476_f00001_fr1_00a00_C_32_Booting PS2 BIOS  _20220901235528 gs xz_accline

PS2
ps2-bios-frame-1-noalpha.bmp

**Monopoly Party_SLES-51145_In_Game.gs.xz**

Some of the corners of the UI boxes look cleaners but others have more gaps.

Master
S003834_00272_f00001_fr1_00000_C_32_Monopoly Party_SLES-51145_In_Game gs xz_master

PR
S003834_00272_f00001_fr1_00000_C_32_Monopoly Party_SLES-51145_In_Game gs xz_accline

PS2
monopoly-frame-1-noalpha

**Ridge Racer V_SLUS-20002_20240613085131.gs.xz**

Master has striped pattern shifted up by 1 compared to PS2/PR due to rounding for lines. PR is missing the first line due to the scissoring optimization in the vertex kick not accounting for line rounding (will do in a separate PR).

Master
S095319_00088_f00002_fr1_00000_C_32_Ridge Racer V_SLUS-20002_20240613085131 gs xz_master

PR
S095319_00088_f00002_fr1_00000_C_32_Ridge Racer V_SLUS-20002_20240613085131 gs xz_accline

PS2
ps2-ridge-racer-frame-2-noalpha

**VP2Bloom.gs.xz**

PR/Master HW/PS2 seems to miss one line in the minimap due to different line rounding. Not sure what is causing this because the RTs appears to be fine when the line is actually drawn. Might have something to with the postprocessing bilinear filter used for antialiasing.

Master SW
S001375_00231_f00001_fr1_01000_C_24_VP2Bloom gs xz_master

Master HW
00231_f00001_fr1_01000_C_24_master_gl

PR SW
S001375_00231_f00001_fr1_01000_C_24_VP2Bloom gs xz_accline

PS2
[ps2-VP2Bloom2-frame-1-noalpha.bmp

**golf.gs.xz**

PR is missing one of the lines in power-up bar. This is a different issue related to https://github.com/PCSX2/pcsx2/pull/12975. The bbox will need to be adjusted again to be slightly larger to account for line rounding (will do in a separate PR).

Master
S000557_04496_f00002_fr1_00a00_C_32_golf gs xz_master

PR
S000557_04496_f00002_fr1_00a00_C_32_golf gs xz_accline

### Details of the pixel inclusion

**Lines**

For aliased lines, the inclusion is the standard method of rounding to nearest pixel in y direction for shallow lines and x direction for steep lines. This was mostly correct in master other than coordinates half way between two pixels should be rounded up instead of floored as currently done.

The criteria for determining whether first/last pixel in the line is drawn is a diamond exit condition (sort of mentioned in the manual; see [here](https://learn.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-rasterizer-stage-rules#line-rasterization-rules-aliased-without-multisampling) also). The diamond region depends on the whether the line is shallow (right-to-left or left-to-right) or steep (up-to-down or down-to-up). For shallow left-to-right line, the diamond contains the top-right border, excludes the right corner, excludes the bottom-right border, and includes the bottom and top corners (see diagram below). The last pixel is drawn only if the diamond is entered and exited; the first pixels is drawn in the opposite cases (makes sense to prevent overlap in a line strip).

diamond_edit

(Above) The diamond region for left-to-right shallow lines. For right-to-left, the region will be flipped horizontally. For steep lines the region will be rotated/flipped accordingly.

For antialiased, it's pretty much what the manual says: instead of rounding to the nearest pixel, consider the two nearest pixels and give then coverage proportational to distance to the line. If the line is exactly on a pixel center, there are still two pixels included: one with full coverage and another with no coverage. In this case the pixel with no coverage is the one in the bottom or right direction. Combined with accurate rounding and first/last pixel inclusion from above, this appears to given accurate antialiased lines (up to errors with calculating the alpha as mentioned earlier).

### Triangles

Aliased triangles already seems to be totally accurate. Possibly there are rounding error somewhere since we use floating point slope, though I did not test it fully.

For antialiased triangles, the procedure is basically what the manual says: treat each edge of the triangle as a line antialias it. However, instead of the including pixels on both sides of the line, just include the one outside the triangles. The antialised pixel must also be on the inside of the other two edges. So if we have a triangle with edges A, B, C, and we are antialiasing edge A, all the antialiased pixels should be on the *outside* of A and the *inside* of B and C (see figure below).

triangle_edit

(Above) The red dots indicate the pixels that will be considered part of the antialias region for edge A. The green dots would be candidated but they are not on the correct side of C and B.

The bounds of the antialiased line seem to be determined by flooring/ceiling the two vertices of the edge and sometimes adding +/-1 to extend the bound. For example, a shallow edge with vertices (x0,y0) -- (x1,y1) might allow the antialiased pixels to range from floor(x0) to ceil(x0). However, in other cases it seems to behave like ceil(x0) - 1 to floor(x1) + 1. I could not determine the exact rule; may require more research.

Some quirks:
1. A top or left edge of a triangle that passes through pixel center can give rise to 0 coverage pixel outside the triangle.
2. A bottom or right edge of a triangle that that passes through pixel center will only give rise to full coverage pixels on the edge (since pixels on bottom or right edges are omitted from the interior of the triangle by the GS).

A different but related issue is that Currently GSState::VertexKick() culls triangles that cover 0 pixel. This is correct for aliased triangles but not for antialiased triangles so may require a separate PR to resolve.

## Other notes

* As mentioned above, alpha calculation based on coverage seems a bit quirky. Sometimes pixels that should have full coverage get alpha 127 (instead of 128). More research may be required.
* Interpolation of vertex attributes is still not done totally accurately for triangles. Right now we treat antialiased edges as a line and simply interpolate the attributes one-dimensionally in the axis being stepped. However, to be more accurate I believe we would need to do full two-dimensional interpolation. Since the gradients for both x and y are already computed in triangle setup it might not add too much overhead to do this.
* Expect a (hopefully small) performance drop in line drawing and antialiased triangle for the extra overhead needed to accuracy. Hopefully, line drawing and antialiased triangle edge drawing is minor enough that it doesn't affect overall performance too much. With that said, the code it probably not as optimized as it could be so I would appreciate any suggestions on this front.

### CODE REVIEW

**Comment by TJnotJT** [comment]
Small addition: Had overlooked point rounding but they are also rounded up to nearest pixel center.

File: pcsx2/GS/Renderers/SW/GSRasterizer.cpp [hunk 1]
```diff
@@ -296,108 +296,352 @@ void GSRasterizer::DrawPoint(const GSVertexSW* vertex, int vertex_count, const u
...
+			e22 = e2 + (step_x ? efun2.y : efun2.x) * offset;
+		};
 
-			scan.p = v0.p.blend32(v1.p, mask);
-			scan.t = v0.t.blend32(v1.t, mask);
-			scan.c = v0.c.blend32(v1.c, mask);
+		if (d > 0.0f)
+		{
+			cov = static_cast<int>(0xffff * (side ? 1.0 - d : d));
+			constexpr int offset = (side ? 0 : 1);
```
Comment (lightningterror) [inline pcsx2/GS/Renderers/SW/GSRasterizer.cpp]: Wonder if marking it `[[maybe_unused]]` will silence codacy, same for the other warnings.

Comment (TJnotJT) [reply pcsx2/GS/Renderers/SW/GSRasterizer.cpp]: Good call, that seems to have resolved a couple of them.

For the remaining, I just removed the assertion. Code analyzer was right that it was redundant though it was kind of for future-proofing -- not a big deal though.

**Comment by lightningterror** [comment]
toto_aa dump looks odd, dunno how it's supposed to look on ps2 so that's worth checking
Master:
image
PR:
image

**Comment by TJnotJT** [comment]
Good catch, this is an actual bug. Here is PS2 for comparison:
ps2-toto-aa-frame-2-noalpha

There are small (e.g., 1 pixel) antialiased lines being draw where the prestep extrapolates alpha to a negative number, which rolls over to a large value and gets blended. An easy fix is to just clamp the colors, it might be good to do a few hardware tests to roughly find out what PS2 is doing there. To be honest, I didn't really consider attribute interpolation carefully as that is a whole other subject, though of course we don't want regressions.

Another related issue is that it seems that previously there was no prestep for aliased lines, only antialiased. So need to find out what is the correct behavior here.

Converting to draft until this is ironed out.

**Comment by TJnotJT** [comment]
After hardware tests it looks like colors are clamped as well as Z but not UVs. Similarly, I will assume that the fog coefficient is clamped but STQs are not. It's not totally clear where in the pipeline the clamping actually happens so this might not be totally accurate. However, it fixes the off-pixels in the dump above.

Also, it appears from the tests that both aliased and antialiased lines have a prestep.

Not fully tested so leaving as draft.

**Comment by TJnotJT** [comment]
Dump in SW mode looks good so removing draft status.

**Review by JordanTheToaster — APPROVED** [review]
Working and looking fine performance in some specific cases is a bit lower but overall performance seems unchanged.
